Adding PredicateChoice to Paper API (updated version)

CPieter opened the pull request:
This PR is an updated version of derverdox's PredicateChoice PR #9996. Since the branch in the old PR was quite outdated I reïmplemented the changes in a new branch, but most of the changes remain the same. Credit of the code should go to derverdox and Machine-Maker.
The feature seems to work fine from my testing but I'm unfamiliar with the Paper API so I please let me know if I missed something.

CPieter committed:
add PredicateChoice

lynxplay commented on paper-api/src/main/java/org/bukkit/inventory/RecipeChoice.java:
```diff
@@ -290,4 +291,50 @@ public String toString() {
         // Paper end - check valid ingredients
     }
+
+    // Paper start - add PredicateChoice
```
No need for paper comments in the API anymore.

lynxplay commented on paper-api/src/main/java/org/bukkit/inventory/RecipeChoice.java:
```diff
@@ -290,4 +291,50 @@ public String toString() {
+        public PredicateChoice(Predicate<ItemStack> predicate, ItemStack exampleStack) {
+            Preconditions.checkArgument(predicate != null, "The item predicate cannot be null");
+            Preconditions.checkArgument(exampleStack != null, "The example stack cannot be null");
+            Preconditions.checkArgument(!exampleStack.getType().isAir(), "Cannot have empty/air example stack");
```
Why not just !exampleStack.isEmpty() ?

lynxplay commented on paper-server/patches/sources/net/minecraft/world/item/crafting/Ingredient.java.patch:
```diff
@@ -26,10 +27,15 @@
++        if (this.stackPredicate != null) {
++            return this.stackPredicate.test(stack);
++        }
++        // Paper end
```
Missing ` - Add PredicateChoice`

lynxplay commented on paper-server/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java:
```diff
@@ -32,7 +35,14 @@ static Ingredient toIngredient(RecipeChoice bukkit, boolean requireNotEmpty) {
             stack = Ingredient.of();
-        } else if (bukkit instanceof RecipeChoice.MaterialChoice) {
+        }
+        // Paper start - add PredicateChoice
```
No more comments, can be properly formatted with the above closing if bracket.

lynxplay reviewed (changes requested):
Welcome to paper :tada: 
Thank you for updating the PR. I don'T have time for a indepth review rn, but I left some initial easy nitpicks around comments :+1: 

Will get to a review next week probably

CPieter committed:
fixup code and paper comments

CPieter committed:
add Nullable annotation

marked stackPredicate field in Ingredient as Nullable

CPieter committed:
Update PredicateChoice

Renamed field predicate to stackPredicate

CPieter committed:
Implement recipe book matching

Machine-Maker commented on paper-api/src/main/java/org/bukkit/inventory/RecipeChoice.java:
```diff
@@ -290,4 +291,48 @@ public String toString() {
+     * given predicate
+     */
+    @NullMarked
+    public static class PredicateChoice implements RecipeChoice {
```
We'd like to avoid having classes/records or really any implementation types exposed to the API, preferring to just have an interface.

So make this an interface, and add a package-private implementation, in the api, like `PredicateChoiceImpl` that is a record. Then inside `RecipeChoice`, add a static method to create an instance of it.

CPieter replied:
What's your opinion on also adding static constructors inside `RecipeChoice` for the `MaterialChoice` and `ExactChoice`? (for concistency)

Machine-Maker replied:
Yeah, let’s do that. Then mark the constructors for those as deprecated. We’ll move those to interfaces too when we have the bytecode tools to support rewriting that.

Machine-Maker replied:
Call the method “itemTypeChoice”.

CPieter replied:
Is there even a point to having seperate interfaces? Maybe these methods could all just return a RecipeChoice.

Machine-Maker replied:
might be too big of a change. There are specific public methods for each.

Actually just ignore MaterialChoice for now. I'm gonna open a separate PR handling that one, just do ExactChoice

Machine-Maker commented on paper-api/src/main/java/org/bukkit/inventory/RecipeChoice.java:
```diff
@@ -290,4 +291,48 @@ public String toString() {
+            Preconditions.checkArgument(!exampleStack.isEmpty(), "Cannot have empty/air example stack");
+
+            this.stackPredicate = stackPredicate;
+            this.exampleStack = exampleStack;
```
In the constructor of the above-mentioned record, clone the itemstack input.

CPieter committed:
Replaced RecipeChoice class with interface+record

CPieter committed:
Added static constructors for RecipeChoice subtypes

Added static constructor methods for MaterialChoice and ExactChoice to the RecipeChoice interface.
Deprecated old constructors.

CPieter committed:
Added Javadocs and moved argument checks to impl

CPieter committed:
Edit Javadoc

CPieter committed:
Removed static MaterialChoice constructors

CPieter committed:
Update StackedContentsExtrasMap

Changed if to else-if, to prevent PredicateChoice recipe book examples from being added as exactIngredient

Machine-Maker commented on paper-api/src/main/java/org/bukkit/inventory/PredicateChoiceImpl.java:
```diff
@@ -0,0 +1,38 @@
+
+        this.stackPredicate = stackPredicate;
+        this.exampleStack = exampleStack.clone();
+    }
```
```suggestion
    public PredicateChoiceImpl {
        Preconditions.checkArgument(stackPredicate != null, "The item predicate cannot be null");
        Preconditions.checkArgument(exampleStack != null, "The example stack cannot be null");
        Preconditions.checkArgument(!exampleStack.isEmpty(), "Cannot have empty/air example stack");
        exampleStack = exampleStack.clone()
    }
```
can use a fancy record compact constructor here since its just preconditions

Machine-Maker commented on paper-api/src/main/java/org/bukkit/inventory/RecipeChoice.java:
```diff
@@ -290,4 +343,14 @@ public String toString() {
+    @NullMarked
+    sealed interface PredicateChoice extends RecipeChoice permits PredicateChoiceImpl {
+
+        Predicate<ItemStack> getPredicate();
```
Not sure we have to expose this, since it is a predicate itself. An empty interface is fine. To your point about maybe not needing interfaces, I think it's good to have even an empty one so you can at least know what type of choice it is. otherwise it's a little annoying to figure out.

Machine-Maker commented on paper-server/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java:
You also have to handle the conversion in the other direction. When it converts from an nms ingredient to a bukkit RecipeChoice. Check if the predicate is set and create one there. It's just below the method you changed.

Machine-Maker commented on paper-api/src/main/java/org/bukkit/inventory/RecipeChoice.java:
```diff
@@ -79,10 +121,12 @@ public MaterialChoice(@NotNull Material... choices) {
             this(new ArrayList<>(java.util.Objects.requireNonNull(choices, "Cannot create a material choice with null tag").getValues())); // Paper - delegate to list ctor to make sure all checks are called
         }
 
+        @Deprecated
```
Lets not deprecate since we don't have the alternative in, and Material isn't deprecated yet.

Machine-Maker commented on paper-api/src/main/java/org/bukkit/inventory/RecipeChoice.java:
```diff
@@ -192,14 +236,23 @@ public static class ExactChoice implements RecipeChoice {
+        /**
+         * @deprecated Use {@link RecipeChoice#exactChoice(List)} instead
+         */
         public ExactChoice(@NotNull List<ItemStack> choices) {
```
add `@Deprecated(since = "1.21.4", forRemoval = true)` to all of these.

Machine-Maker commented on paper-api/src/main/java/org/bukkit/inventory/RecipeChoice.java:
```diff
@@ -35,6 +37,46 @@ public interface RecipeChoice extends Predicate<ItemStack>, Cloneable {
+     *               Cannot be empty or contain empty/air stacks.
+     * @return a new ExactChoice.
+     */
+    static @NotNull ExactChoice exactChoice(@NotNull ItemStack... stacks) {
```
Lets avoid upstream's mistake. For situations where it's not valid to pass 0 values, and you want a varargs parameter, you can do smth like.
```suggestion
    static @NotNull ExactChoice exactChoice(@NotNull ItemStack first, ItemStack... others) {
```
That forces them to supply at least one value, and optionally anymore

Machine-Maker reviewed (changes requested):
Looks good, just a few more things. I also [opened the PR](https://github.com/PaperMC/Paper/pull/12049) adding the future alternative to MaterialChoice. Will merge mine in after yours

CPieter committed:
Small fixes

CPieter committed:
Small Javadoc and contract changes

CPieter committed:
Added NotNull annotation

Machine-Maker committed:
switch over potionmix

Doc94 commented on paper-api/src/main/java/org/bukkit/inventory/PredicateRecipeChoiceImpl.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public RecipeChoice validate(final boolean allowEmptyRecipes) {
+        if (this.exampleStack.getType().isAir()) {
```
```suggestion
        if (this.exampleStack.isEmpty()) {
```

kennytv replied:
The constructor already checks against this, so the check can be removed fully

kennytv commented on paper-server/patches/features/0020-Improve-exact-choice-recipe-ingredients.patch:
```diff
@@ -108,7 +112,9 @@ index 0000000000000000000000000000000000000000..f47c12e9dd6cfa857ca07a764edc22de
 +    public void initialize(final Recipe<?> recipe) {
 +        this.exactIngredients.clear();
```
Doesn't the new list have to be cleared here as well?

kennytv commented on paper-api/src/main/java/org/bukkit/inventory/PredicateRecipeChoiceImpl.java:
```diff
@@ -0,0 +1,41 @@
+    @SuppressWarnings({"MethodDoesntCallSuperMethod", "FunctionalExpressionCanBeFolded"})
+    @Override
+    public PredicateRecipeChoiceImpl clone() {
+        return new PredicateRecipeChoiceImpl(this.stackPredicate::test, this.exampleStack.clone());
```
The constructor already clones
